Add `.to()` to OVBaseModel

helena-intel opened the pull request:
Fixes `.to()` not working for OVModelForCausalLM

helena-intel committed:
Add `.to()` to OVBaseModel

Fixes `.to()` not working for OVModelForCausalLM

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-intel/pr_284). All of your documentation changes will be reflected on that endpoint.

helena-intel committed:
Moved .to() to OVBaseDecoderModel

Vipitis commented:
Hey,
I don't think this sufficiently fixes #279 as OVBaseCoderModel constructs a `model.decoder` [OVDecoder](https://github.com/huggingface/optimum-intel/blob/ab18f01c136aaa1ac98a129abe8c8b54f372ddcf/optimum/intel/openvino/modeling_decoder.py#LL107C12-L107C12) with `_devive`. And `.to()` does only change  `model._device` not `model.decoder._device` which is used for `.compile()` in OVBaseDecoderModel

helena-intel commented:
The simple implementation of moving to OVBaseModel silently failed. After `.to("GPU")` inference did not occur on GPU but still on CPU, because the decoder model was not moved to the new device. For now I moved the existing implementation back to OVModel and added `.to()` to OVBaseDecoderModel and explicitly set the correct device to the decoder model there.

helena-intel commented:
@Vipitis Thanks! We wrote at the same time. I noticed it too and it should be fixed now.

helena-intel commented:
@echarlaix the test fails with PyTorch 4.28. I restricted PyTorch to <4.28 on my fork and the tests pass there: https://github.com/helena-intel/optimum-intel/actions/runs/4699850769/jobs/8333817513

echarlaix reviewed (commented):
Thanks a lot for fixing it @helena-intel

echarlaix commented:
> @echarlaix the test fails with PyTorch 4.28 (unrelated to this PR). I restricted PyTorch to <4.28 on my fork and the tests pass there: https://github.com/helena-intel/optimum-intel/actions/runs/4699850769/jobs/8333817513

Yes you're correct it's unrelated to this PR, I fixed it in https://github.com/huggingface/optimum-intel/pull/285

helena-intel commented on optimum/intel/openvino/modeling_decoder.py:
```diff
@@ -208,6 +208,16 @@ def reshape(self, batch_size: int, sequence_length: int):
+        """
+        self._device = device.upper()
+        self.decoder._device = device.upper()
+        self.request = None
```
Thanks @echarlaix !

helena-intel committed:
Update optimum/intel/openvino/modeling_decoder.py

Co-authored-by: Ella Charlaix <80481427+echarlaix@users.noreply.github.com>
